projects/gstreamer/gst-discoverer.c

```diff
@@ -0,0 +1,505 @@
+  GstDiscovererStreamInfo *sinfo;
+
+  if (!info) {
+    g_print ("Could not discover URI\n");
+    g_print (" %s\n", err->message);
+    return;
+  }
+
+  result = gst_discoverer_info_get_result (info);
+  g_print ("Done discovering %s\n", gst_discoverer_info_get_uri (info));
+  switch (result) {
+    case GST_DISCOVERER_OK:
+    {
+      break;
+    }
+    case GST_DISCOVERER_URI_INVALID:
+    {
+      g_print ("URI is not valid\n");
+      break;
+    }
+    case GST_DISCOVERER_ERROR:
+    {
+      g_print ("An error was encountered while discovering the file\n");
+      g_print (" %s\n", err->message);
+      break;
+    }
+    case GST_DISCOVERER_TIMEOUT:
+    {
+      g_print ("Analyzing URI timed out\n");
+      break;
+    }
+    case GST_DISCOVERER_BUSY:
+    {
+      g_print ("Discoverer was busy\n");
+      break;
+    }
+    case GST_DISCOVERER_MISSING_PLUGINS:
+    {
+      g_print ("Missing plugins\n");
+      if (verbose) {
+        gint i = 0;
+        const gchar **installer_details =
+            gst_discoverer_info_get_missing_elements_installer_details (info);
+
+        while (installer_details[i]) {
+          g_print (" (%s)\n", installer_details[i]);
+
+          i++;
+        }
+      }
+      break;
+    }
+  }
+
+  if ((sinfo = gst_discoverer_info_get_stream_info (info))) {
+    g_print ("\nTopology:\n");
+    print_topology (sinfo, 1);
+    g_print ("\nProperties:\n");
+    print_properties (info, 1);
+    gst_discoverer_stream_info_unref (sinfo);
+  }
+
+  g_print ("\n");
+}
+
+const guint8 *fuzztesting_data;
+size_t fuzztesting_size;
+
+static void
+appsrc_configuration (GstDiscoverer *dc, GstElement *source, gpointer data)
+{
+  GstBuffer *buf;
+  GstFlowReturn ret;
+  
+  /* Create buffer from fuzztesting_data */
+  buf = gst_buffer_new_wrapped ((gpointer) fuzztesting_data, fuzztesting_size);
+  g_object_set (G_OBJECT (source), "size", fuzztesting_size, NULL);
+  g_signal_emit_by_name (G_OBJECT(source), "push-buffer", buf, &ret);
+}
+
+int LLVMFuzzerTestOneInput(const guint8 *data, size_t size)
+{
+  GError *err = NULL;
+  GstDiscoverer *dc;
+  gint timeout = 10;
+  GstDiscovererInfo *info;
+
+  gst_init (NULL, NULL);
+  
+  GST_PLUGIN_STATIC_REGISTER(coreelements);
+  GST_PLUGIN_STATIC_REGISTER(playback);
+  GST_PLUGIN_STATIC_REGISTER(typefindfunctions);
+  GST_PLUGIN_STATIC_REGISTER(app);
+  GST_PLUGIN_STATIC_REGISTER(ogg);
+  GST_PLUGIN_STATIC_REGISTER(theora);
+  GST_PLUGIN_STATIC_REGISTER(vorbis);
+  
+  dc = gst_discoverer_new (timeout * GST_SECOND, &err);
+  if (G_UNLIKELY (dc == NULL)) {
+    g_print ("Error initializing: %s\n", err->message);
+    g_clear_error (&err);
+    exit (1);
+  }
+
+  fuzztesting_data = data;
+  fuzztesting_size = size;
+  
+  /* Connect to source-setup signal to give the data */
+  g_signal_connect (dc, "source-setup", (GCallback) appsrc_configuration, NULL);
+  
+  info = gst_discoverer_discover_uri (dc, "appsrc://", &err);
+  print_info (info, err);
```

Comment:
If you print something while executing every input, fuzzing will be very inefficient (will spend all the time printing)

Reply:
Is one/two lines ok for debugging purposes ?

Reply:
I think you may have it under some debug switch, in case you are going to debug things locally, but for fuzzing build we need to avoid any output.

Reply:
Fair enough, I'll fix that.